src/ray/core_worker/store_provider/plasma_store_provider.cc

```diff
@@ -16,72 +16,101 @@ Status CoreWorkerPlasmaStoreProvider::Put(const RayObject &object,
   return local_store_provider_.Put(object, object_id);
 }
 
+Status CoreWorkerPlasmaStoreProvider::Create(const std::shared_ptr<Buffer> &metadata,
+		                               const size_t data_size,
+                                               const ObjectID &object_id,
+					       std::shared_ptr<Buffer> &data) {
+  return local_store_provider_.Create(metadata, data_size, object_id, data);
+}
+
+Status CoreWorkerPlasmaStoreProvider::Seal(const ObjectID &object_id) {
+  return local_store_provider_.Seal(object_id);
+}
+
 Status CoreWorkerPlasmaStoreProvider::Get(
     const std::vector<ObjectID> &ids, int64_t timeout_ms, const TaskID &task_id,
     std::vector<std::shared_ptr<RayObject>> *results) {
   (*results).resize(ids.size(), nullptr);
-
-  bool was_blocked = false;
-
-  std::unordered_map<ObjectID, int> unready;
+  RAY_CHECK_OK(raylet_client_->FetchOrReconstruct(ids, /*fetch_only=*/true, task_id));
+  RAY_RETURN_NOT_OK(local_store_provider_.Get(ids, 0, task_id, results));
+
+  // Iterate through the results from the local store, adding them to the unready
+  // map if they weren't successfully fetched from the local store (are nullptr).
+  // Keeps track of the locations of the unready object IDs in the original list
+  // (accounting for duplicates).
+  std::unordered_map<ObjectID, std::vector<int>> unready;
   for (size_t i = 0; i < ids.size(); i++) {
-    unready.insert({ids[i], i});
+    if ((*results)[i] != nullptr) {
+      continue;
+    }
+    auto it = unready.find(ids[i]);
+    if (it == unready.end()) {
+      std::vector<int> v;
+      v.push_back(i);
+      unready.insert({ids[i], v});
+    } else {
+      it->second.push_back(i);
+    }
   }
 
+  // If all objects were fetched in the first batch, return.
+  if (unready.empty()) {
+    return Status::OK();
+  }
+
+  // If not all objects were successfully fetched, repeatedly call FetchOrReconstruct and 
+  // Get from the local object store in batches. This loop will run indefinitely until the
+  // objects are all fetched if timeout is -1.
   int num_attempts = 0;
   bool should_break = false;
   int64_t remaining_timeout = timeout_ms;
-  // Repeat until we get all objects.
+  int64_t batch_size = RayConfig::instance().worker_fetch_request_size();
   while (!unready.empty() && !should_break) {
     std::vector<ObjectID> unready_ids;
     for (const auto &entry : unready) {
+      if (unready_ids.size() == batch_size) {
+        break;
+      }
```

Comment:
It seems that you'll only `FetchOrReconstruct` one batch here?  I think we should reconstruct all objects in parallel.

Reply:
`FetchOrReconstruct` can block the raylet if we call it with a very large list of objects, which is the reason for the batching. Note that the batch size is large by default (10000), so the common case is to FetchOrReconstruct everything in parallel.

Reply:
I see. however, I'm afraid this change makes this function too complex. It'd be nice to unify the code for the first attempt and the remaining attempts as much as possible.

Reply:
actually, I think there is another issue here.  For example, if we're getting 10 batches of objects. And objects will be ready in 1s. Previously, this get operation can finish in 1s. But now, we'll need 10s (sleeping 10 times).

Considering that getting more than 10k objects is a very rare case, it's doesn't seem to be worth it to sacrifice code readability and introduce the above-mentioned problem. I'd suggest reverting this change from this PR. What do you think?